src/bin/pgcopydb/cli_list.c

```diff
@@ -1252,72 +1240,79 @@ cli_list_table_parts(int argc, char **argv)
 
 	Catalogs *catalogs = &(copySpecs.catalogs);
 	DatabaseCatalog *sourceDB = &(catalogs->source);
-	SourceTable *table = (SourceTable *) calloc(1, sizeof(SourceTable));
 
-	if (table == NULL)
+	ListTablePartFetchPartsContext context = { .sourceDB = sourceDB };
```

Comment:
I could have fetched all table names and their partition using this simple query
```
  select nspname, relname, partnum, partcount, min, max, count
                    from s_table_part parts
         join s_table t on parts.oid = t.oid;
```

but that would require adding new catalog functions. I believe it's not worth it for a utility command like this. So leveraged the existing `catalog_iter_s_table` with the callback calling `catalog_iter_s_table_parts` for each table.  The memory cost wouldn't be too different and compute performance isn't significant either.

Reply:
Well we could have a million tables in the catalogs, but only a few dozens large enough to use COPY partitioning, so I believe it's worth adding the extra catalog query. You might be able to re-use an existing iterator and just change the init function where there is the SQL query, so that you only introduce 2 new functions: the main iterator and the associated iterator init function.